spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java

```diff
@@ -792,11 +793,14 @@ void setMessageBuilderFactory(MessageBuilderFactory messageBuilderFactory) {
 		}
 
 		@Override
-		public Message<?> toMessage(Object object) throws Exception {
+		public Message<?> toMessage(Object object, @Nullable Map<String, Object> headers) throws Exception {
 			if (object instanceof Message<?>) {
 				return (Message<?>) object;
 			}
-			return (object != null) ? this.messageBuilderFactory.withPayload(object).build() : null;
+
+			return object != null
+					? this.messageBuilderFactory.withPayload(object).copyHeadersIfAbsent(headers).build()
```

Comment:
We should add `@Nullable` to `copyHeadersIfAbsent` (it's present on the Spring Messaging MB).

Reply:
Is it related to this PR ?

There are too much those Kotlin "love" in SF...

Reply:
Yes, it's related because you now unconditionally `copyHeadersIfAbsent(headers)` - I had to go and check the code to see if it's protected. `@Nullable` documents it.

Reply:
??? That was always unconditional because there is `if (headersToCopy != null) {`.

I don't understand your concern regarding this `@Nullable`. 
This is something new in SF `5.0` and we have lived without it for a while already.

What is the point to jumt into this subject now?
Thanks

Reply:
? This cHIA () is new code in this commit and the arg is @Nullable.

Reply:
How is that new? 
I just use here existing `MessageBuilder` API.
What am I missing?

Right, I added a new `headers` arg to the `toMessage()` method and just followed its interface contract with adding `@Nullable`. And this is definitely new code. But why do I have to go from here everywhere to fix all the possible `@Nullable`?

Cant't we address it on demand at least?

Reply:
You don't have to "go everywhere". Since you can send null here I had to check if it was safe. So this is indeed "on demand". This discussion is more effort than it would have taken ro fix it.